File: Regression/Polynomial-Regression.R
# Polynomial Linear Regression

# importing dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]
#dataset1 = dataset[,2:3]

# fitting linear regression model to the dataset
lin_reg = lm(formula = Salary ~ .,data = dataset)

# fitting polynomial regression model to dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., data = dataset)

# visualising linear regression model
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)), color = 'blue') +
  ggtitle('Level vs Salary') +
  xlab('Level') +
  ylab('Salary')

# visualising polynomial regression model
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)), color = 'blue')+
  ggtitle('Level vs Salary') +
  xlab('Level') +
  ylab('Salary')

# Visualising the Regression Model results (for higher resolution and smoother curve)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), color = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg, newdata = data.frame(Level = x_grid,
                                                                              Level2 = x_grid^2,
                                                                              Level3 = x_grid^3,
                                                                              Level4 = x_grid^4))),
                color = 'blue')+
  ggtitle('Level vs Salary') +
  xlab('Level') +
  ylab('Salary')

# predicting values using linear regression
y_pred_linear = predict(lin_reg, newdata = data.frame(Level = 6.5))

# predicting values using polynomial regression
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))

File: Regression/Polynomial-Regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 24 15:14:28 2017

@author: adhingra
"""

# Polynmial Linear Regression

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# importing dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:,1:2].values   # to make matrix, provide upper bound
y = dataset.iloc[:,2].values

# we won't be splitting dataset into test and training sets as the obersvations are a few

# fitting linear regression model to dataset
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg.fit(X,y)

# fitting polynomial regression model to dataset
from sklearn.preprocessing import PolynomialFeatures
polyreg = PolynomialFeatures(degree=4)
X_poly = polyreg.fit_transform(X)
linreg2 = LinearRegression()
linreg2.fit(X_poly,y)

# visualising linear regression results
plt.scatter(X,y,color='red')
plt.plot(X, linreg.predict(X), color = 'blue')
plt.title('Level vs Salary - Linear Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# visualising polynomial regression results
plt.scatter(X,y,color='red')
plt.plot(X,linreg2.predict(polyreg.fit_transform(X)),color='blue')
plt.title('Level vs Salary - Polynomial Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# graph smoothening and higher resolution for visualising polynomial regression
X_grid = np.arange(min(X),max(X),0.1)
X_grid = X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color='red')
plt.plot(X_grid,linreg2.predict(polyreg.fit_transform(X_grid)),color = 'blue')
plt.title('Level vs Salary - Polynomial Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# predicitng a new result with linear regrssion
y_pred_lin = linreg.predict(6.5)

# predicting a new result with polynomial regression
y_pred_poly = linreg2.predict(polyreg.fit_transform(6.5))


